### PR TITLE
add extensible members in protocol.

### PR DESCRIPTION
**Pull request by vwxyzh**
Allow adding more optional members in future.

## Issue

Serialize and deserialize optional members.

## Scenario

We will have more optional members in near future.

* Tracing Id -  for diagnosis
* Sequence Id - for ordering in some special time (client connection switch to another server connection in the same server).

We cannot use current length based deserializing to detect which field occurred.

## Solution
Add a map field represent all optional fields in the binary.

* All of optional fields should write in this field.
* Non-optional fields are not recommended in this field.
* Key is field hint for identifying the optional field (e.g.: tracing id is 1).
* Value is the real value for the optional field.

### Sample
The original format for `ConnectionDataMessage` is:
```
[(type)6, (ConnectionId) "abc", (payload) binary]
```
Updated format is:
```
[(type)6, (ConnectionId) "abc", (payload) binary, (optional fields) { }]
```
And with tracing id `id123`:
```
[(type)6, (ConnectionId) "abc", (payload) binary, (optional fields) { (hint)1: (value)"id123" }]
```

## About this PR

This PR just want to standardize the optional fields serializing/deserializing.
Define optional fields or implement them is out of scope.

### CODE REVIEW

**Comment by vicancy** [comment]
There is already one in the way https://github.com/Azure/azure-signalr/pull/820

I also have a requirement to add some additional `Metadata` field

File: src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs [hunk 1]
```diff
@@ -12,6 +14,49 @@ public abstract class ServiceMessage
...
     }
 
+    /// <summary>
+    /// Base class of messages between Azure SignalR Service and SDK.
+    /// </summary>
+    public abstract class ExtensibleServiceMessage : ServiceMessage
+    {
+        private const int TracingId = 1;
+
+        internal void WriteExtensionMembers(ref MessagePackWriter writer)
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs:24]: This method isn't used

File: src/Microsoft.Azure.SignalR.Protocols/IMessageWithTracingId.cs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Microsoft.Azure.SignalR.Protocol
+{
+    public interface IMessageWithTracingId
+    {
+        string TracingId { get; set; }
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Protocols/IMessageWithTracingId.cs:5]: How about call it `Id` instead?

File: src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs [hunk 3]
```diff
@@ -12,6 +14,49 @@ public abstract class ServiceMessage
...
+            }
+            writer.WriteArrayHeader(count * 2);
+            if (!string.IsNullOrEmpty(tracingId))
+            {
+                writer.Write(TracingId);
+                writer.Write(tracingId);
+            }
+        }
+
+        internal void ReadExtensionMembers(ref MessagePackReader reader)
```
Comment (wanlwanl) [inline src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs:40]: This method isn't used

Comment (vwxyzh) [reply src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs:40]: pls extend diff in serviceprotocol.cs

File: src/Microsoft.Azure.SignalR.Protocols/IMessageWithTracingId.cs (hunk 2 above)
Comment (vwxyzh) [reply src/Microsoft.Azure.SignalR.Protocols/IMessageWithTracingId.cs:5]: pls extend diff in serviceprotocol.cs

Comment (vwxyzh) [reply src/Microsoft.Azure.SignalR.Protocols/IMessageWithTracingId.cs:5]: just a sample, you can change it.

File: src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs [hunk 4]
```diff
@@ -12,6 +14,49 @@ public abstract class ServiceMessage
     {
     }
 
+    /// <summary>
+    /// Base class of messages between Azure SignalR Service and SDK.
+    /// </summary>
+    public abstract class ExtensibleServiceMessage : ServiceMessage
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs:20]: what if I only need extensible to `ConnectionDataMessage`? just have `ExtensibleServiceMessage` to `ServiceMessage` as TraceId sounds applicable for all the messages?

Comment (vwxyzh) [reply src/Microsoft.Azure.SignalR.Protocols/ServiceMessage.cs:20]: yes, leave it to @wanlwanl 😃

**Comment by davidfowl** [comment]
What's this for exactly? Why isn't it extensible today? Is there a spec on this design?

**Comment by vwxyzh** [comment]
@davidfowl I add more detials for this pr.

**Comment by davidfowl** [comment]
1. Do all new fields go into the extension field?
2. How does it affect the size of the message?

**Comment by vwxyzh** [comment]
@davidfowl 
1. In my proposal, only optional fields should go into the extension field.
    If the field is required, I do not suggest, although I don't think there will be a new required field.
2. Message size: all kinds of messages allowing optional fields will append one more byte (an empty map) at least.
   And for messages have optional field, one more integer value (bytes based on the value of integer, -31 to 127 takes one byte) for hint, and the bytes for value itself.

**Pull request merged**